// @api-platform/admin/lib/stories/Basic.stories.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import React from 'react';
import { within } from '@storybook/test';
import { HydraAdmin } from '../hydra';
import { OpenApiAdmin } from '../openapi';
import DevtoolsLayout from './layout/DevtoolsLayout';
/**
 * # Basic `<HydraAdmin>` component
 * The `<HydraAdmin>` component without any parameter.
 */
const Basic = ({ entrypoint }) => (React.createElement(HydraAdmin, { entrypoint: entrypoint, layout: DevtoolsLayout }));
const meta = {
    title: 'Admin/Basic',
    component: Basic,
    parameters: {
        layout: 'fullscreen',
    },
};
export default meta;
export const Hydra = {
    play: (_a) => __awaiter(void 0, [_a], void 0, function* ({ canvasElement }) {
        const canvas = within(canvasElement);
        yield canvas.findByText('Greetings');
    }),
    args: {
        entrypoint: process.env.ENTRYPOINT,
    },
};
export const OpenApi = () => (React.createElement(OpenApiAdmin, { entrypoint: process.env.ENTRYPOINT, docEntrypoint: `${process.env.ENTRYPOINT}/docs.jsonopenapi` }));
